File: public/js/models/game.js
var Game = function(){
  this.board = (function(){
    var nested_board = [];
    for(var i = 0; i < 7; i++)
    {
      var column = [];
      for(var j = 0; j < 6; j++)
      {
        column.push(0);
      }
      nested_board.push(column);
    }
    return nested_board;
  })();
}

Game.prototype.gameOver = function(){
  var col_results = evaluateLinear(this.board)
  if(col_results === false)
  {
    var row_results = evaluateLinear(transpose(this.board))
    if(row_results === false)
    {
      var diagonal = evaluateDiag(this.board)
      if(diagonal === false)
      {
        nested = this.board.slice(0);
        nested.reverse();
        return evaluateDiag(nested);
      }
      else
        return diagonal
    }
    else
     return row_results
  }
  else
    return col_results
}
Game.prototype.addPiece = function(data){
  var column = data.columnId
  var player = data.player
  var changed = false;
  for(i=0; i<this.board[column].length; i++ )
  {
    if(this.board[column][i] !== 0)
      {
        changed = true;
        this.board[column][i-1] = player;
      }
  }

}
function evaluateDiag(nested_array)
{
  var nested_diags = [];
  for (var k =3; k <9; k++ )
  {
    var diagonal = [];
    for(var i = 0; i < nested_array.length; i++)
    {
      for(var j = 0; j < nested_array[i].length; j++)
      {
        if((i + j) === k)
        {
          diagonal.push(nested_array[i][j])
        }
      }
    }
    nested_diags.push(diagonal);
  }
  return evaluateLinear(nested_diags);
}

function evaluateLinear(nested_array)
{
   for(var i=0; i<nested_array.length; i++)
   {
    var counter = 0;
    var value;
    for(var j=0; j<nested_array[i].length; j++)
    {
      if (nested_array[i][j] === 0) {
        counter = 0;
      }
      else if (value === nested_array[i][j])
      {
        counter += 1
      }
      else
      {
        value = nested_array[i][j]
        counter = 1;
      }
      if(counter === 4)
        return value;
    }
   }
   return false;
}
function transpose(a)
{
  return a[0].map(function (_, c) { return a.map(function (r) { return r[c]; }); });
}

File: app/models/game.rb
class Game < ActiveRecord::Base
  # Remember to create a migration!
  @@current_player = 1

  def self.addPiece(column)
    column = column.to_i - 1
    col = "col" + column.to_s
    col.to_sym
    row = Game.where("#{col} != ?", 0)
    if row.length == 0
      last_row = Game.last
      eval "last_row.#{col} = #{@@current_player}"
      last_row.save
    else
      eval "row.#{col} = #{@@current_player}"
      row.save
    end
    self.switch_player
  end

  def self.switch_player
    @@current_player = 1 if @@current_player = 2
    @@current_player = 2 if @@current_player = 1
  end

  def self.gameOver
    gameboard = Game.all.as_json.to_a
    nested_game = []
    gameboard.each do |col|
      col.delete("id")
      nested_game << col.values
    end
    col_results = evaluateLinear(nested_game)
    if !col_results
      return evaluateLinear(nested_game.transpose)
    else
     return col_results

  end

  def self.evaluateLinear(board)
    board.each do |col|
      counter = 0
      player = nil
      col.delete("id")
      col.values.each do |element|
        if element == 0
          counter = 0
        elsif player == element
          counter +=1
        else
          player = element
          counter = 1
        end
        if counter == 4
          return player
        end
      end
    end
    return false
  end

  def self.reset
    Game.all.each do |row|
      row.col0 = 0;
      row.col1 = 0;
      row.col2 = 0;
      row.col3 = 0;
      row.col4 = 0;
      row.col5 = 0;
      row.col6 = 0;
    end
  end
end
